akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -0,0 +1,296 @@
+
+      case v @ AliveConnectionTerminators(ts) ⇒
+        if (terminators.compareAndSet(v, Terminating(timeout.fromNow))) {
+          // cause the termination for all connections
+          val connectionsTerminated = Future.sequence(ts.map { t ⇒
+            t.terminate(timeout)
+          })
+          val serverTerminated = connectionsTerminated.map(_ ⇒ HttpServerTerminated)
+          termination.completeWith(serverTerminated)
+          termination.future
+        } else terminate(timeout)(ex) // retry
+
+      case Terminating(existingDeadline) ⇒
+        log.warning(s"Issued terminate($timeout) while termination is in progress already (with deadline: time left: ${PrettyDuration.format(existingDeadline.timeLeft)}")
+        termination.future
+    }
+  }
+}
+
+final class ServerTerminationDeadlineReached()
+  extends RuntimeException("Server termination deadline reached, shutting down all connections and terminating server...")
+
+object GracefulTerminatorStage {
+  def apply(system: ActorSystem, serverSettings: ServerSettings): BidiFlow[HttpResponse, HttpResponse, HttpRequest, HttpRequest, ServerTerminator] = {
+    val stage = new GracefulTerminatorStage(serverSettings)
+    BidiFlow.fromGraph(stage)
+  }
+}
+
+/**
+ * See detailed docs termination process on [[akka.http.scaladsl.Http.ServerBinding]].
+ *
+ * Stage shape diagram:
+ *
+ * {{{
+ *                      +---+
+ * fromNet Request   -> | G | -> toUser Request
+ *                      | T |
+ * toNet   Response  <- | S | <- fromUser Response
+ *                      +---+
+ * }}}
+ */
+final class GracefulTerminatorStage(settings: ServerSettings)
+  extends GraphStageWithMaterializedValue[BidiShape[HttpResponse, HttpResponse, HttpRequest, HttpRequest], ServerTerminator] {
+
+  val fromNet: Inlet[HttpRequest] = Inlet("netIn")
+  val toUser: Outlet[HttpRequest] = Outlet("userOut")
+
+  val fromUser: Inlet[HttpResponse] = Inlet("userIn")
+  val toNet: Outlet[HttpResponse] = Outlet("netOut")
+  override def shape = BidiShape(fromUser, toNet, fromNet, toUser)
+
+  final val TerminationDeadlineTimerKey = "TerminationDeadlineTimerKey"
+
+  final class ConnectionTerminator(triggerTermination: Promise[FiniteDuration ⇒ Future[HttpTerminated]]) extends ServerTerminator {
+    override def terminate(deadline: FiniteDuration)(implicit ec: ExecutionContext): Future[HttpTerminated] = {
+      triggerTermination.future.flatMap(callback ⇒ {
+        callback(deadline)
+      })
+    }
+  }
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ServerTerminator) = {
+    val triggerTermination = Promise[FiniteDuration ⇒ Future[HttpTerminated]]() // result here means termination of this connection has completed
+
+    // responsible for terminating this connection
+    val selfTerminator = new ConnectionTerminator(triggerTermination)
+
+    val logic = new TimerGraphStageLogic(shape) with StageLogging {
```

Comment:
good to override `logSource` for better logger name

Reply:
good :)